src/lib.rs

```diff
@@ -66,17 +89,112 @@ impl VehiclePropertyValue {
         }
     }
 
-    fn check_valid_type(&self, vhal_prop_type: &VehiclePropertyType) -> bool {
+    fn from_prop(
+        prop_type: &VehiclePropertyType,
+        prop_value: &VehicleHalProto::VehiclePropValue,
+    ) -> Result<Self> {
+        Ok(match *prop_type {
+            VehiclePropertyType::STRING => Self::String(
+                prop_value
+                    .string_value
+                    .as_ref()
+                    .ok_or(VhalError::PropertyTypeError)?
+                    .to_owned(),
+            ),
+            VehiclePropertyType::BYTES => Self::Bytes(
+                prop_value
+                    .bytes_value
+                    .as_ref()
+                    .ok_or(VhalError::PropertyTypeError)?
+                    .to_vec(),
+            ),
+            VehiclePropertyType::BOOLEAN => Self::Int32(
+                *prop_value
+                    .int32_values
+                    .first()
+                    .ok_or(VhalError::PropertyTypeError)?,
+            ),
+            VehiclePropertyType::INT32 => Self::Int32(
+                *prop_value
+                    .int32_values
+                    .first()
+                    .ok_or(VhalError::PropertyTypeError)?,
+            ),
+            VehiclePropertyType::INT64 => Self::Int64(
+                *prop_value
+                    .int64_values
+                    .first()
+                    .ok_or(VhalError::PropertyTypeError)?,
+            ),
+            VehiclePropertyType::FLOAT => Self::Float(
+                *prop_value
+                    .float_values
+                    .first()
+                    .ok_or(VhalError::PropertyTypeError)?,
+            ),
+            _ => return Err(VhalError::ReceiveMessageValueTypeError),
+        })
+    }
+
+    fn check_valid_type(&self, prop_type: &VehiclePropertyType) -> bool {
         match self {
-            Self::String(_) => vhal_prop_type.is_string(),
-            Self::Int32(_) => vhal_prop_type.is_int32(),
-            Self::Int64(_) => vhal_prop_type.is_int64(),
-            Self::Float(_) => vhal_prop_type.is_float(),
-            Self::Bytes(_) => vhal_prop_type.is_bytes(),
+            Self::String(_) => prop_type.is_string(),
+            Self::Int32(_) => prop_type.is_int32(),
+            Self::Int64(_) => prop_type.is_int64(),
+            Self::Float(_) => prop_type.is_float(),
+            Self::Bytes(_) => prop_type.is_bytes(),
         }
     }
 }
 
+impl GetPropertyValue for EmulatorMessage {
+    fn get_value(&self) -> Result<VehiclePropertyValue> {
+        if self.value.len() != 1 {
```

Comment:
If it can't be zero, maybe better change to `>1`.

Reply:
No, the only valid value is 1. With `>1` it would allow empty vectors.

Reply:
No, you misunderstood me.
I meant the following:
`!=1` means either 0 or more than 1 (2, 3, ...)
If zero is possible returning `ReceiveMessageTooManyValuesError` is confusing.
If zero of values is not possible, then changing `!=1` into `>1` is more readable.

Reply:
But never mind, this is not that important, just explaining myself :)

Reply:
Ah! gotcha. Well, at least the size of the values vector is printed with the error :D